domain/secretbackend/state/state.go

```diff
@@ -512,51 +521,86 @@ WHERE uuid = $M.uuid`, sqlair.M{}, SecretBackendRotationRow{})
 	return domain.CoerceError(err)
 }
 
+// GetModelSecretBackend returns the secret backend name for the specified model.
+func (s *State) GetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID) (string, coremodel.ModelType, error) {
+	db, err := s.DB()
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+	var modelBackendDetails domainsecretbackend.ModelSecretBackend
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		modelBackendDetails, err = s.getModelSecretBackendDetails(ctx, modelUUID, tx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return "", "", domain.CoerceError(err)
+	}
+	return modelBackendDetails.SecretBackendName, modelBackendDetails.Type, nil
+}
+
 // SetModelSecretBackend sets the secret backend for the given model.
-func (s *State) SetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID, backendName string) error {
+func (s *State) SetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID, getBackendName func(coremodel.ModelType) (string, error)) error {
 	db, err := s.DB()
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	modelBackendUpsert := `
-INSERT INTO model_secret_backend
-	(model_uuid, secret_backend_uuid)
-VALUES ($M.model_uuid, $SecretBackendRow.uuid)
-ON CONFLICT (model_uuid) DO UPDATE SET
-	secret_backend_uuid = EXCLUDED.secret_backend_uuid`
-	modelBackendUpsertStmt, err := s.Prepare(modelBackendUpsert, sqlair.M{}, SecretBackendRow{})
+	secretBackendSelectQ := `
+SELECT b.uuid AS &ModelSecretBackend.secret_backend_uuid
+FROM   secret_backend b
+WHERE  b.name = $ModelSecretBackend.secret_backend_name
+`
+	backendInfo := ModelSecretBackend{ID: modelUUID}
+	secretBackendSelectStmt, err := s.Prepare(secretBackendSelectQ, backendInfo)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	q := `
-SELECT b.uuid AS &SecretBackendRow.uuid
-FROM secret_backend b
-WHERE b.name = $SecretBackendRow.name`
-	stmt, err := s.Prepare(q, SecretBackendRow{})
+	modelBackendUpdate := `
+UPDATE model_secret_backend
+SET    secret_backend_uuid = $ModelSecretBackend.secret_backend_uuid
+WHERE  model_uuid = $ModelSecretBackend.uuid`
+	modelBackendUpdateStmt, err := s.Prepare(modelBackendUpdate, backendInfo)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		backend := SecretBackendRow{Name: backendName}
+		modelBackendDetails, err := s.getModelSecretBackendDetails(ctx, modelUUID, tx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		backendInfo.SecretBackendName, err = getBackendName(modelBackendDetails.Type)
```

Comment:
As per comments above we should never ever be doing this. Aside from the layer boundaries we are violating this call has the potential to cripple a whole controller with no way out.

We are inside of a transaction hold it open while we made indeterminate call to a closure that we have with no context passed to stop what is going on. Please remove this logic and get it back up into the services layer. The decision on what the backend is should be well and truly made before we get anywhere near a transaction or the state layer.